File: examples/form-comparison/servers/javascript/records.mjs
// The JavaScript record server (docs/spec/form-comparison.md, "Record resource"): one persistent
// store of the customer records, its HTTP contract and the SSR stage views, rendered with the
// JavaScript generators and checked with the JavaScript validator. The public server answers
// `/api/js/records…` with this module and `servers/javascript/main.mjs` serves it on its own.
import { randomBytes } from 'node:crypto';
import { readFile, rename, rm, writeFile } from 'node:fs/promises';
import path from 'node:path';

import { compileForm, createForm } from '@crudui/generator-core';
import { renderDetail, renderForm, renderList } from '@crudui/generator-html';
import { FormInputError, validate } from '@crudui/validator';

import {
  formData, linkedSpecs, recordClients, recordLanguages, recordModes, recordsPerPage, recordViews,
} from '../../src/record-view.mjs';

const bodyLimit = 2 * 1024 * 1024;
const positiveInteger = /^[1-9][0-9]*$/;
// The submitted form (docs/spec/form-comparison.md, "Record resource"): `text` and `checkbox`
// leaves, `texts`, an object of exactly these text members, `fields`, an object of these fields,
// and `rows`, keyed rows of one object of fields.
const text = 'text';
const checkbox = 'checkbox';
const texts = (...members) => ({ texts: members });
const fields = members => ({ fields: members });
const rows = members => ({ rows: fields(members) });
const required = shape => ({ required: shape });
const rowKey = /^__[0-9a-f]{13}__$/;
const formShape = fields({
  id: required(text), name: text, status: text, joined: text, score: text, relation: texts('name'), markup: text,
  companies: rows({
    name: text,
    stores: rows({
      name: text, enabled: checkbox, detail: text, title: texts('ko', 'en'),
      departments: rows({ name: text }),
    }),
  }),
});

class HttpError extends Error {
  constructor(status, message) {
    super(message);
    this.status = status;
  }
}

const isObject = value => value !== null && typeof value === 'object' && !Array.isArray(value);

/** The value of an absent field: empty text, empty texts or no rows. */
function empty(shape) {
  if (shape === text || shape === checkbox) return '';
  if (shape.texts) return Object.fromEntries(shape.texts.map(member => [member, '']));
  return {};
}

/**
 * The submitted `value` of `shape` at `path`, completed and in the member order of the shape.
 * Form data leaves out a field that holds no value, so an absent field other than a required one
 * completes as its empty value in both media types. Any other difference answers 400.
 */
function shaped(value, shape, path) {
  if (shape === text || shape === checkbox) {
    if (typeof value !== 'string') throw new HttpError(400, `Expected text at ${path}`);
    if (shape === checkbox && value !== '' && value !== '1') throw new HttpError(400, `Expected "" or "1" at ${path}`);
    return value;
  }
  if (!isObject(value)) throw new HttpError(400, `Expected an object at ${path}`);
  const result = {};
  if (shape.rows) {
    for (const [key, row] of Object.entries(value)) {
      if (!rowKey.test(key)) throw new HttpError(400, `Expected a row key at ${path}: ${key}`);
      result[key] = shaped(row, shape.rows, `${path}.${key}`);
    }
    return result;
  }
  const names = shape.texts ?? Object.keys(shape.fields);
  const extra = Object.keys(value).find(member => !names.includes(member));
  if (extra !== undefined) throw new HttpError(400, `Unexpected member ${path}.${extra}`);
  for (const member of names) {
    // An object of texts holds every member; a field may be absent unless it is required.
    const child = shape.texts ? required(text) : shape.fields[member];
    const field = child.required ?? child;
    if (Object.hasOwn(value, member)) result[member] = shaped(value[member], field, `${path}.${member}`);
    else if (child.required) throw new HttpError(400, `Missing member ${path}.${member}`);
    else result[member] = empty(field);
  }
  return result;
}

// The members of a stored record, in order (docs/spec/form-comparison.md, "Record resource", "Store").
const recordMembers = ['id', 'name', 'status', 'joined', 'score', 'relation', 'avatar', 'markup', 'companies'];

/**
 * Require the records array of a store or of the fixture: records with exactly the fixture's
 * members in order, `score` a number, every other scalar a string and `companies` complete in the
 * form's shape and member order. Anything else is a server fault.
 */
function checkRecords(value, name) {
  const malformed = (detail, cause) => new Error(`The record ${name} is malformed: ${detail}`, { cause });
  if (!Array.isArray(value)) throw malformed('expected a records array');
  for (const record of value) {
    if (!isObject(record) || Object.keys(record).join() !== recordMembers.join()) throw malformed('expected the record members');
    for (const member of recordMembers) {
      const item = record[member];
      const valid = member === 'score' ? typeof item === 'number'
        : member === 'relation' ? isObject(item) && Object.keys(item).join() === 'name' && typeof item.name === 'string'
          : member === 'companies' ? true : typeof item === 'string';
      if (!valid) throw malformed(`expected the record member ${member}`);
    }
    let companies;
    try {
      companies = shaped(record.companies, formShape.fields.companies, 'companies');
    } catch (error) {
      if (!(error instanceof HttpError)) throw error;
      throw malformed(error.message, error);
    }
    // A stored record holds its companies complete, in member order.
    if (JSON.stringify(companies) !== JSON.stringify(record.companies)) throw malformed('expected complete companies in member order');
  }
  return value;
}

/** The submitted object of native fields `form[…]…` with `_form_complete=1`. */
function nativeForm(fields) {
  // Objects without a prototype keep every field name, `__proto__` included, an own member.
  const form = Object.create(null);
  let complete = 0;
  for (const [name, value] of fields) {
    if (typeof value !== 'string') throw new HttpError(400, `Field ${name} must be text`);
    if (name === '_form_complete') {
      if (value !== '1' || ++complete > 1) throw new HttpError(400, 'Expected one _form_complete=1');
      continue;
    }
    const match = /^form((?:\[[^[\]]+\])+)$/.exec(name);
    if (!match) throw new HttpError(400, `Unknown field ${name}`);
    const segments = match[1].slice(1, -1).split('][');
    const last = segments.pop();
    let parent = form;
    for (const segment of segments) {
      if (!Object.hasOwn(parent, segment)) parent[segment] = Object.create(null);
      else if (typeof parent[segment] !== 'object') throw new HttpError(400, `Repeated field ${name}`);
      parent = parent[segment];
    }
    if (Object.hasOwn(parent, last)) throw new HttpError(400, `Repeated field ${name}`);
    parent[last] = value;
  }
  if (complete !== 1) throw new HttpError(400, 'The form submission is incomplete');
  return form;
}

/**
 * Read one request body up to 2 MiB. A body over the limit fails with 413 as soon as it passes the
 * limit: the rest is never read, and the response closes the connection.
 */
function readBody(request) {
  return new Promise((resolve, reject) => {
    const chunks = [];
    let size = 0;
    const settle = (action, value) => {
      request.off('data', onData);
      request.off('end', onEnd);
      request.off('error', onError);
      request.pause();
      action(value);
    };
    const onData = chunk => {
      size += chunk.length;
      if (size > bodyLimit) settle(reject, new HttpError(413, 'Request exceeds 2 MiB'));
      else chunks.push(chunk);
    };
    const onEnd = () => settle(resolve, Buffer.concat(chunks));
    const onError = error => settle(reject, error);
    request.on('data', onData);
    request.once('end', onEnd);
    request.once('error', onError);
  });
}

/** Parse one save request as the rendered form or as JSON posts it. */
async function parseSubmission(request, body) {
  const type = (request.headers['content-type'] ?? '').split(';')[0].trim().toLowerCase();
  if (type === 'application/json') {
    let value;
    try {
      value = JSON.parse(body.toString('utf8'));
    } catch {
      throw new HttpError(400, 'Malformed JSON');
    }
    if (!isObject(value) || Object.keys(value).length !== 1 || !Object.hasOwn(value, 'form')) {
      throw new HttpError(400, 'Expected { "form": { … } }');
    }
    return shaped(value.form, formShape, 'form');
  }
  if (type !== 'multipart/form-data' && type !== 'application/x-www-form-urlencoded') {
    throw new HttpError(415, 'Expected multipart/form-data, application/x-www-form-urlencoded or application/json');
  }
  let fields;
  try {
    fields = await new Request('http://localhost/', {
      method: 'POST', headers: { 'Content-Type': request.headers['content-type'] }, body,
    }).formData();
  } catch {
    throw new HttpError(400, 'Malformed form request');
  }
  return shaped(nativeForm(fields), formShape, 'form');
}

/** Parse a view query: `id` first for detail and form, then the selection, each exactly once. */
function viewSelection(view, search, server) {
  const entries = [...new URLSearchParams(search)];
  const expected = [...(view === 'list' ? [] : ['id']), 'lang', 'server', 'framework', 'initialization', 'mode', 'page'];
  if (entries.length !== expected.length || entries.some(([key], index) => key !== expected[index])) {
    throw new HttpError(400, `Expected the query ${expected.join(', ')}`);
  }
  const query = Object.fromEntries(entries);
  if (!recordLanguages.includes(query.lang) || query.server !== server || !recordClients.includes(query.framework)
      || query.initialization !== 'ssr' || !recordModes.includes(query.mode) || !positiveInteger.test(query.page)
      || (view !== 'list' && !positiveInteger.test(query.id))) {
    throw new HttpError(400, 'Invalid view query');
  }
  return { id: query.id, selection: { ...query, id: undefined, page: Number(query.page) } };
}

/**
 * The record store of one server in `dataDirectory`, reading the published fixture and
 * specifications from `publicDirectory`. Writes run one at a time and replace the file atomically.
 */
export function recordStore({ server = 'js', dataDirectory, publicDirectory }) {
  const storeFile = path.join(dataDirectory, `records-${server}.json`);
  const readPublished = async name => JSON.parse(await readFile(path.join(publicDirectory, name), 'utf8'));
  const readFixture = async () => checkRecords(await readPublished('customer-records.json'), 'fixture');
  let queue = Promise.resolve();

  /** Run `task` after every earlier store operation of this process. */
  function exclusive(task) {
    const result = queue.then(task);
    queue = result.catch(() => {});
    return result;
  }

  async function write(records) {
    const temporary = path.join(dataDirectory, `.records-${server}-${randomBytes(6).toString('hex')}.tmp`);
    try {
      await writeFile(temporary, JSON.stringify(records, null, 2) + '\n');
      await rename(temporary, storeFile);
    } finally {
      await rm(temporary, { force: true });
    }
  }

  async function read() {
    let text;
    try {
      text = await readFile(storeFile, 'utf8');
    } catch (error) {
      if (error.code !== 'ENOENT') throw error;
      const records = await readFixture();
      await write(records);
      return records;
    }
    let records;
    try {
      records = JSON.parse(text);
    } catch (error) {
      throw new Error(`The record store is malformed: ${error.message}`, { cause: error });
    }
    return checkRecords(records, 'store');
  }

  const records = () => exclusive(read);
  const lastPage = total => Math.max(1, Math.ceil(total / recordsPerPage));
  const pageData = (stored, page) => {
    if (page > lastPage(stored.length)) throw new HttpError(404, 'Page not found');
    return {
      page, perPage: recordsPerPage, total: stored.length,
      records: stored.slice((page - 1) * recordsPerPage, page * recordsPerPage),
    };
  };
  const findRecord = (stored, id) => {
    const record = stored.find(item => item.id === id);
    if (!record) throw new HttpError(404, 'Record not found');
    return record;
  };

  /** One page of records, or a 404 failure beyond the last page. */
  async function list(page) {
    return pageData(await records(), page);
  }

  /** One stored record, or a 404 failure. */
  async function record(id) {
    return { record: findRecord(await records(), id) };
  }

  /** The SSR stage of one view for a view query string. */
  async function view(name, search) {
    if (!recordViews.includes(name)) throw new HttpError(404, 'Unknown view');
    const { id, selection } = viewSelection(name, search, server);
    const [stored, published] = await Promise.all([records(), readPublished('customer-specs.json')]);
    const specs = linkedSpecs(published, selection);
    const language = selection.lang;
    if (name === 'list') {
      const data = pageData(stored, selection.page);
      const html = renderList(specs.list, data.records, { language, layout: 'table', page: data.page, total: data.total });
      return { view: name, html, data };
    }
    const item = findRecord(stored, id);
    if (name === 'detail') {
      return { view: name, html: renderDetail(specs.detail, item, { language }), data: { record: item } };
    }
    const template = compileForm(specs.form, { keyPrefix: 'form' });
    const form = renderForm(createForm(template, formData(item), { language }));
    const html = `<form id="record-form" method="post" action="/api/${server}/records/${item.id}" `
      + `enctype="multipart/form-data">${form}</form>`;
    return { view: name, html, data: { record: item } };
  }

  async function save(id, request) {
    const body = await readBody(request);
    const specs = await readPublished('customer-specs.json');
    return exclusive(async () => {
      const stored = await read();
      const index = stored.findIndex(item => item.id === id);
      if (index < 0) throw new HttpError(404, 'Record not found');
      const form = await parseSubmission(request, body);
      if (form.id !== id) throw new HttpError(400, 'The submitted id differs from the record id');
      let validation;
      try {
        validation = validate(specs.form, form);
      } catch (error) {
        if (error instanceof FormInputError) throw new HttpError(400, error.message);
        throw error;
      }
      if (!validation.valid) return { status: 422, body: { validation } };
      const previous = stored[index];
      const saved = {};
      for (const [member, value] of Object.entries(previous)) {
        saved[member] = member === 'id' || member === 'avatar' ? value
          : member === 'score' ? Number(form.score)
            : member === 'relation' ? { name: form.relation.name } : form[member];
      }
      stored[index] = saved;
      await write(stored);
      return { status: 200, body: { record: saved, validation: { valid: true, errors: [] } } };
    });
  }

  async function reset(request) {
    const body = await readBody(request);
    if (body.length > 0) throw new HttpError(400, 'A reset takes no request body');
    const fixture = await readFixture();
    await exclusive(() => write(fixture));
    return { total: fixture.length };
  }

  /** Route one request on the server's own paths (`/api/records…`). */
  async function route(method, pathname, search, request) {
    const allow = allowed => {
      if (!allowed.includes(method)) throw new HttpError(405, 'Method not allowed');
    };
    if (pathname === '/api/records') {
      allow(['GET']);
      const entries = [...new URLSearchParams(search)];
      if (entries.length !== 1 || entries[0][0] !== 'page' || !positiveInteger.test(entries[0][1])) {
        throw new HttpError(400, 'Expected one page parameter');
      }
      return { status: 200, body: await list(Number(entries[0][1])) };
    }
    if (pathname === '/api/records/reset') {
      allow(['POST']);
      return { status: 200, body: await reset(request) };
    }
    const viewMatch = /^\/api\/records\/view\/([^/]+)$/.exec(pathname);
    if (viewMatch) {
      allow(['GET']);
      return { status: 200, body: await view(viewMatch[1], search) };
    }
    const idMatch = /^\/api\/records\/([^/]+)$/.exec(pathname);
    if (idMatch) {
      allow(['GET', 'POST']);
      if (method === 'POST') return save(idMatch[1], request);
      return { status: 200, body: await record(idMatch[1]) };
    }
    throw new HttpError(404, 'Unknown endpoint');
  }

  /**
   * Answer one request whose path is already relative to the server (`/api/records…`). Every
   * response is JSON that names this server.
   */
  async function handle(request, response, pathname, search) {
    let result;
    try {
      result = await route(request.method, pathname, search, request);
    } catch (error) {
      const status = error instanceof HttpError ? error.status : 500;
      result = { status, body: { error: error.message } };
    }
    // A body that was not read to its end is not read further: the connection closes.
    const connection = request.complete ? {} : { Connection: 'close' };
    response.writeHead(result.status, {
      'Content-Type': 'application/json; charset=utf-8', 'Cache-Control': 'no-store', ...connection,
    });
    response.end(JSON.stringify({ ...result.body, server }));
  }

  /**
   * Call one read of the store in process, with the status the HTTP contract gives it:
   * `list(page)`, `record(id)` or `view(name, search)`.
   */
  async function call(action, ...args) {
    try {
      return { status: 200, json: { ...await ({ list, record, view })[action](...args), server } };
    } catch (error) {
      if (!(error instanceof HttpError)) throw error;
      return { status: error.status, json: { error: error.message, server } };
    }
  }

  return { handle, call };
}

File: examples/form-comparison/servers/go/records.go
package main

// The customer record resource of the canonical page (docs/spec/form-comparison.md, "Record
// resource"): one store file in the data directory, seeded from the fixture in the public
// directory, and list, detail, save, reset and SSR view routes.

import (
	"errors"
	"fmt"
	"io"
	"math"
	"mime"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"regexp"
	"slices"
	"strconv"
	"strings"
	"unicode"

	generator "github.com/polyspec/crudui/packages/generator-go"
	"github.com/polyspec/crudui/packages/validator-go/validator/validate"
)

const recordsPerPage = 20

// recordMembers is the member order of a stored record; formMembers are the submitted members.
var (
	recordMembers = []string{"id", "name", "status", "joined", "score", "relation", "avatar", "markup"}
	formMembers   = []string{"id", "name", "status", "joined", "score", "relation", "markup"}
	positiveText  = regexp.MustCompile(`^[1-9][0-9]*$`)
	recordViews   = []string{"list", "detail", "form"}
	selectionKeys = []string{"lang", "server", "framework", "initialization", "mode", "page"}
	selectionSets = map[string][]string{
		"lang":           {"ko", "en"},
		"server":         {"go"},
		"framework":      {"html", "react", "vue", "svelte"},
		"initialization": {"ssr"},
		"mode":           {"bindForm", "createForm"},
	}
)

// statusError is a failure with its HTTP status.
type statusError struct {
	status  int
	message string
}

func (e statusError) Error() string { return e.message }

func fail(status int, message string) error { return statusError{status, message} }

func recordFailure(w http.ResponseWriter, err error) {
	var known statusError
	if errors.As(err, &known) {
		failure(w, known.status, err)
		return
	}
	failure(w, http.StatusInternalServerError, err)
}

func (s server) recordStore() string { return filepath.Join(s.dataDir, "records-go.json") }

// fixtureRecords reads the seeded records from the public directory.
func (s server) fixtureRecords() ([]any, error) {
	encoded, err := os.ReadFile(filepath.Join(s.publicDir, "customer-records.json"))
	if err != nil {
		return nil, err
	}
	value, err := decodeJSON(encoded)
	if err != nil {
		return nil, err
	}
	return checkRecords(value)
}

// checkRecords requires an array of records with the fixture's members, order and types.
func checkRecords(value any) ([]any, error) {
	records, ok := value.([]any)
	if !ok {
		return nil, fmt.Errorf("Expected a records array")
	}
	for _, item := range records {
		stored, ok := item.(*object)
		if !ok || !slices.Equal(stored.Keys(), recordMembers) {
			return nil, fmt.Errorf("Expected stored record members")
		}
		for _, name := range recordMembers {
			value := get(stored, name)
			switch name {
			case "score":
				_, ok = value.(float64)
			case "relation":
				var relation *object
				if relation, ok = value.(*object); ok {
					_, ok = get(relation, "name").(string)
					ok = ok && slices.Equal(relation.Keys(), []string{"name"})
				}
			default:
				_, ok = value.(string)
			}
			if !ok {
				return nil, fmt.Errorf("Expected stored record member: %s", name)
			}
		}
	}
	return records, nil
}

// storeTransaction runs operation on the stored records under the store lock.
func (s server) storeTransaction(operation func([]any) ([]any, any, error)) (any, error) {
	return lockedFile(s.recordStore(), func() (any, error) { return s.fixtureRecords() },
		func(value any) (any, any, error) {
			records, err := checkRecords(value)
			if err != nil {
				return nil, nil, err
			}
			return operation(records)
		})
}

func (s server) storedRecords() ([]any, error) {
	result, err := s.storeTransaction(func(records []any) ([]any, any, error) { return records, records, nil })
	if err != nil {
		return nil, err
	}
	return result.([]any), nil
}

func recordIndex(records []any, id string) int {
	return slices.IndexFunc(records, func(item any) bool { return get(item.(*object), "id") == id })
}

// pageOf returns the records of one page, or 404 beyond the last page.
func pageOf(records []any, text string) (int, []any, error) {
	last := max(1, (len(records)+recordsPerPage-1)/recordsPerPage)
	page, err := strconv.Atoi(text)
	if err != nil || page < 1 || page > last {
		return 0, nil, fail(http.StatusNotFound, "Page not found")
	}
	first := (page - 1) * recordsPerPage
	return page, records[first:min(first+recordsPerPage, len(records))], nil
}

func (s server) serveRecords(w http.ResponseWriter, r *http.Request) {
	// Path segments are compared as written, without percent-decoding.
	path := r.URL.EscapedPath()
	allowed := []string{http.MethodGet}
	switch {
	case path == "/api/records":
	case path == "/api/records/reset":
		allowed = []string{http.MethodPost}
	case strings.HasPrefix(path, "/api/records/view/"):
	case strings.Count(path, "/") == 3 && path != "/api/records/":
		allowed = []string{http.MethodGet, http.MethodPost}
	default:
		failure(w, http.StatusNotFound, fmt.Errorf("Unknown endpoint"))
		return
	}
	if !slices.Contains(allowed, r.Method) {
		w.Header().Set("Allow", strings.Join(allowed, ", "))
		failure(w, http.StatusMethodNotAllowed, fmt.Errorf("Method not allowed"))
		return
	}
	var response *object
	var err error
	switch {
	case path == "/api/records":
		response, err = s.listRecords(r.URL.RawQuery)
	case path == "/api/records/reset":
		response, err = s.resetRecords(r)
	case strings.HasPrefix(path, "/api/records/view/"):
		response, err = s.recordView(strings.TrimPrefix(path, "/api/records/view/"), r.URL.RawQuery)
	case r.Method == http.MethodGet:
		response, err = s.recordByID(strings.TrimPrefix(path, "/api/records/"))
	default:
		response, err = s.saveRecord(w, r, strings.TrimPrefix(path, "/api/records/"))
	}
	if err != nil {
		recordFailure(w, err)
		return
	}
	status := http.StatusOK
	if validation, ok := get(response, "validation").(*object); ok && get(validation, "valid") == false {
		status = http.StatusUnprocessableEntity
	}
	writeJSON(w, status, response)
}

func (s server) listRecords(rawQuery string) (*object, error) {
	query, err := url.ParseQuery(rawQuery)
	if err != nil || len(query) != 1 || len(query["page"]) != 1 || !positiveText.MatchString(query.Get("page")) {
		return nil, fail(http.StatusBadRequest, "Expected one page parameter")
	}
	records, err := s.storedRecords()
	if err != nil {
		return nil, err
	}
	page, items, err := pageOf(records, query.Get("page"))
	if err != nil {
		return nil, err
	}
	return record("page", page, "perPage", recordsPerPage, "total", len(records), "records", items), nil
}

func (s server) recordByID(id string) (*object, error) {
	records, err := s.storedRecords()
	if err != nil {
		return nil, err
	}
	index := recordIndex(records, id)
	if index < 0 {
		return nil, fail(http.StatusNotFound, "Record not found")
	}
	return record("record", records[index]), nil
}

func (s server) resetRecords(r *http.Request) (*object, error) {
	body, err := io.ReadAll(io.LimitReader(r.Body, 1))
	if err != nil || len(body) != 0 || r.ContentLength > 0 {
		return nil, fail(http.StatusBadRequest, "Reset takes no request body")
	}
	fixture, err := s.fixtureRecords()
	if err != nil {
		return nil, err
	}
	if _, err = s.storeTransaction(func([]any) ([]any, any, error) { return fixture, nil, nil }); err != nil {
		return nil, err
	}
	return record("total", len(fixture)), nil
}

// submittedForm reads the submitted form object of a native or JSON save request.
func submittedForm(w http.ResponseWriter, r *http.Request) (*object, error) {
	kind, _, err := mime.ParseMediaType(r.Header.Get("Content-Type"))
	native := kind == "multipart/form-data" || kind == "application/x-www-form-urlencoded"
	if err != nil || !native && kind != "application/json" {
		return nil, fail(http.StatusUnsupportedMediaType, "Expected a form or JSON request")
	}
	body, err := io.ReadAll(http.MaxBytesReader(w, r.Body, 2*1024*1024))
	var tooLarge *http.MaxBytesError
	if errors.As(err, &tooLarge) {
		return nil, fail(http.StatusRequestEntityTooLarge, "Request body exceeds 2 MiB")
	}
	if err != nil {
		return nil, fail(http.StatusBadRequest, err.Error())
	}
	var root *object
	expected := []string{"form"}
	if native {
		root, err = parseNative(body, r.Header.Get("Content-Type"))
		expected = []string{"form", "_form_complete"}
	} else {
		var value any
		value, err = decodeJSON(body)
		root, _ = value.(*object)
	}
	if err != nil || root == nil {
		return nil, fail(http.StatusBadRequest, "Expected a form request object")
	}
	if native && get(root, "_form_complete") != "1" {
		return nil, fail(http.StatusBadRequest, "Incomplete native form submission")
	}
	if !sameMembers(root.Keys(), expected) {
		return nil, fail(http.StatusBadRequest, "Expected only the form submission fields")
	}
	form, ok := get(root, "form").(*object)
	if !ok || !sameMembers(form.Keys(), formMembers) {
		return nil, fail(http.StatusBadRequest, "Expected the record form members")
	}
	for _, name := range formMembers {
		value := get(form, name)
		if name == "relation" {
			relation, ok := value.(*object)
			if !ok || !sameMembers(relation.Keys(), []string{"name"}) {
				return nil, fail(http.StatusBadRequest, "Expected the relation name member")
			}
			value = get(relation, "name")
		}
		if _, ok := value.(string); !ok {
			return nil, fail(http.StatusBadRequest, "Expected text member: "+name)
		}
	}
	return form, nil
}

func sameMembers(keys, expected []string) bool {
	return len(keys) == len(expected) && !slices.ContainsFunc(keys, func(key string) bool { return !slices.Contains(expected, key) })
}

func (s server) recordSpec(name string) (*object, error) {
	specs, err := readObject(filepath.Join(s.publicDir, "customer-specs.json"))
	if err != nil {
		return nil, err
	}
	spec, ok := get(specs, name).(*object)
	if !ok {
		return nil, fmt.Errorf("Expected the %s specification", name)
	}
	return spec, nil
}

func (s server) saveRecord(w http.ResponseWriter, r *http.Request, id string) (*object, error) {
	records, err := s.storedRecords()
	if err != nil {
		return nil, err
	}
	if recordIndex(records, id) < 0 {
		return nil, fail(http.StatusNotFound, "Record not found")
	}
	form, err := submittedForm(w, r)
	if err != nil {
		return nil, err
	}
	if get(form, "id") != id {
		return nil, fail(http.StatusBadRequest, "The submitted id differs from the record id")
	}
	spec, err := s.recordSpec("form")
	if err != nil {
		return nil, err
	}
	result, err := validate.Validate(spec, validatorData(form), validate.Options{})
	if err != nil {
		return nil, err
	}
	errorList := []any{}
	for _, item := range result.Errors {
		errorList = append(errorList, record("path", item.Path, "field", item.Field, "rule", item.Rule, "message", item.Message, "value", item.Value))
	}
	validation := record("valid", result.Valid, "errors", errorList)
	if !result.Valid {
		return record("validation", validation), nil
	}
	score, err := strconv.ParseFloat(strings.TrimFunc(get(form, "score").(string), func(r rune) bool { return unicode.Is(unicode.White_Space, r) }), 64)
	if err != nil && !errors.Is(err, strconv.ErrRange) || math.IsInf(score, 0) {
		return nil, fail(http.StatusBadRequest, "Expected a finite score")
	}
	saved, err := s.storeTransaction(func(records []any) ([]any, any, error) {
		index := recordIndex(records, id)
		if index < 0 {
			return nil, nil, fail(http.StatusNotFound, "Record not found")
		}
		previous := records[index].(*object)
		next := record()
		for _, name := range recordMembers {
			value := get(form, name)
			switch name {
			case "id", "avatar":
				value = get(previous, name)
			case "score":
				value = score
			case "relation":
				value = record("name", get(get(form, "relation").(*object), "name"))
			}
			next.Set(name, value)
		}
		changed := slices.Clone(records)
		changed[index] = next
		return changed, next, nil
	})
	if err != nil {
		return nil, err
	}
	return record("record", saved, "validation", validation), nil
}

// viewQuery reads the selection of an SSR view request: id first for detail and form, then the
// selection members in order, each once.
func viewQuery(view, rawQuery string) (map[string]string, error) {
	keys := selectionKeys
	if view != "list" {
		keys = append([]string{"id"}, selectionKeys...)
	}
	pairs := strings.Split(rawQuery, "&")
	if rawQuery == "" || len(pairs) != len(keys) {
		return nil, fail(http.StatusBadRequest, "Expected the view selection query")
	}
	values := map[string]string{}
	for index, pair := range pairs {
		rawName, rawValue, _ := strings.Cut(pair, "=")
		name, nameErr := url.QueryUnescape(rawName)
		value, valueErr := url.QueryUnescape(rawValue)
		accepted, listed := selectionSets[name]
		valid := nameErr == nil && valueErr == nil && name == keys[index]
		if listed {
			valid = valid && slices.Contains(accepted, value)
		} else {
			valid = valid && positiveText.MatchString(value)
		}
		if !valid {
			return nil, fail(http.StatusBadRequest, "Expected the view selection query")
		}
		values[name] = value
	}
	return values, nil
}

// selectionText is the selection query every generated link carries.
func selectionText(values map[string]string) string {
	parts := make([]string, len(selectionKeys))
	for index, name := range selectionKeys {
		parts[index] = url.QueryEscape(name) + "=" + url.QueryEscape(values[name])
	}
	return strings.Join(parts, "&")
}

// appendHref appends the selection query to the link format at the given member path.
func appendHref(spec *object, query string, path ...string) error {
	current := spec
	for _, name := range path {
		next, ok := get(current, name).(*object)
		if !ok {
			return fmt.Errorf("Expected specification member: %s", name)
		}
		current = next
	}
	href, ok := get(current, "href").(string)
	if !ok {
		return fmt.Errorf("Expected a link format")
	}
	current.Set("href", href+"&"+query)
	return nil
}

// formText is the form data of one stored record: every member as text, without the avatar.
func formText(stored *object) *object {
	relation := get(stored, "relation").(*object)
	return record("id", get(stored, "id"), "name", get(stored, "name"), "status", get(stored, "status"),
		"joined", get(stored, "joined"), "score", numberText(get(stored, "score").(float64)),
		"relation", record("name", get(relation, "name")), "markup", get(stored, "markup"))
}

func (s server) recordView(view, rawQuery string) (*object, error) {
	if !slices.Contains(recordViews, view) {
		return nil, fail(http.StatusNotFound, "View not found")
	}
	values, err := viewQuery(view, rawQuery)
	if err != nil {
		return nil, err
	}
	records, err := s.storedRecords()
	if err != nil {
		return nil, err
	}
	page, items, err := pageOf(records, values["page"])
	if err != nil {
		return nil, err
	}
	var stored *object
	if view != "list" {
		index := recordIndex(records, values["id"])
		if index < 0 {
			return nil, fail(http.StatusNotFound, "Record not found")
		}
		stored = records[index].(*object)
	}
	spec, err := s.recordSpec(view)
	if err != nil {
		return nil, err
	}
	language, query := values["lang"], selectionText(values)
	var html string
	var data *object
	switch view {
	case "list":
		if err = appendHref(spec, query, "columns", "name", "format"); err != nil {
			return nil, err
		}
		rows := make([]*generator.Object, len(items))
		for index, item := range items {
			rows[index] = item.(*object)
		}
		html, err = generator.RenderList(spec, rows, generator.ListOptions{Language: language, Layout: "table", Page: page, Total: len(records)})
		data = record("page", page, "perPage", recordsPerPage, "total", len(records), "records", items)
	case "detail":
		if err = appendHref(spec, query, "fields", "id", "format"); err != nil {
			return nil, err
		}
		html, err = generator.RenderDetail(spec, stored, generator.DetailOptions{Language: language})
		data = record("record", stored)
	default:
		var template *generator.FormTemplate
		template, err = generator.CompileForm(spec, generator.CompileOptions{KeyPrefix: "form", KeyPrefixProvided: true})
		if err != nil {
			return nil, err
		}
		var form *generator.Form
		form, err = generator.NewForm(template, formText(stored), generator.BindOptions{Language: language})
		if err != nil {
			return nil, err
		}
		html, err = generator.RenderForm(form)
		html = `<form id="record-form" method="post" action="/api/go/records/` + values["id"] +
			`" enctype="multipart/form-data">` + html + `</form>`
		data = record("record", stored)
	}
	if err != nil {
		return nil, err
	}
	return record("view", view, "html", html, "data", data), nil
}
